include/openssl/core_numbers.h

```diff
@@ -112,9 +112,9 @@ OSSL_CORE_MAKE_FUNC(void *, OP_digest_dupctx, (void *dctx))
 OSSL_CORE_MAKE_FUNC(size_t, OP_digest_size, (void))
 OSSL_CORE_MAKE_FUNC(size_t, OP_digest_block_size, (void))
 OSSL_CORE_MAKE_FUNC(int, OP_digest_set_params,
-                    (void *vctx, const OSSL_PARAM params[]))
+                    (void *vctx, OSSL_PARAM params[]))
```

Comment:
There's actually no reason whatsoever to de-constify any `set_params` type of function.  `return_size` will (or should) never be touched.

Reply:
The entire point of this is to move the return type into the OSSL_PARAM structure rather than using a pointer.

Reply:
So? `set_params` functions should still not set it. `get_params` functions are a different matter...

Reply:
I think you're mistaking the two sides of the operations.  The set and get functions are intended for the receiver (non-owner).  When that side sets a value the owner will want to know the length, hence `return_size` is set.  When the receiver gets a value, the `return_size` shouldn't be because nothing is being returned.

Might this be indicative as to the confusing nature....

I don't mind making both the set and get calls put an appropriate value into the `return_size` field.  Perhaps call it `actual_size` or `used_size`.

Reply:
I am not mistaken.  A `set_params` function is used by the `OSSL_PARAM` array *owner* to pass that array.  The `set_params` function itself is the *receiving* end, and will use the `OSSL_PARAM_get` functions to get the values that were passed.

The `set_params` and `get_params` names were chosen for the *owning* side, i.e `set_params` sets parameters, and `get_params` gets parameters.  For the former, I expect the return size to never be used, and for the latter, I expect the return size to always be set.

Reply:
I suggest a closer study of `test/params_test.c`, where this pattern is used.

Reply:
We've got a very different view on this :(

As was pointed out sometime back, the _owner_ never needs to call set or get -- direct access to the value is available.  E.g.
```
int foo;
OSSL_PARAM p = OSSL_PARAM_int("foo", &foo);
```
Why call `OSSL_PARAM_set_int(p, 3);` when `foo = 3;` is better?  Likewise get `OSSL_PARAM_get_int(p, &foo);` is a nop.

The current `OSSL_PARAM_get` calls do not set the `return_size`, the `OSSL_PARAM_set` calls do.

If the receiver sets the value, the return size needs to be passed back to the owner.

Reply:
I think _setter_ and _responder_ are getting confused.  The _responder_ calls the OSSL_PARAM_get functions to read values from the _setter_ and OSSL_PARAM_set to write them so they can be used by the _setter_ on return.

The OSSL_PARAM_[gs]et familiy of functions as from the viewpoint the _responding_ side.

Reply:
@paulidale, will you please look at the line I've made this remark on?

``` C
OSSL_CORE_MAKE_FUNC(int, OP_digest_set_params,
                    (void *vctx, const OSSL_PARAM params[]))
```

That will form this function type:

``` C
typedef int (OSSL_OP_digest_set_params)(void *vctx, const OSSL_PARAM params[]);
```

That's the signature for a *setter*, i.e. a function that helps *set* internal parameters of something, and a function matching that type will typically call functions like `OSSL_PARAM_get_int` using the `OSSL_PARAM` array it has been passed.

In contrast, we have the following:

``` C
OSSL_CORE_MAKE_FUNC(int, OP_digest_get_params,
                    (void *vctx, OSSL_PARAM params[]))
```

That will form this function type:

``` C
typedef int (OSSL_OP_digest_get_params)(void *vctx, OSSL_PARAM params[]);
```

That's the signature for a *responder*, i.e. a function that helps *get* internal parameters of something, and a function matching that type will typically call functions like `OSSL_PARAM_set_int` using the `OSSL_PARAM` array it has been passed.

Please note how I made the *setter* `OSSL_PARAM` const, while I made the *responder* `OSSL_PARAM` non-const.

Now, regarding terminology...  any of the functions that match the signatures are receivers of `OSSL_PARAM` arrays.  However, they are (ultimately) called by an `OSSL_PARAM` array owner, right?  The names of those function types have been chosen to match the expectations of the caller, i.e. the caller is "setting" internal parameters, or the caller is "getting" internal parameters.  The confusing bit is the receiving end seems to do the exact opposite if you only look at the names...  however, it's actually pretty logical if you look at how you'd write a much simpler setter/getter pair:

``` C
int SOMETHING_set_foo(SOMETHING *obj, int value)
{
    /* in OSSL_PARAM terms, this would be OSSL_PARAM_get_int(p, &obj->foo),
       'cause you're getting the new value of obj->foo from |value| */
    obj->foo = value;
}

int SOMETHING_get_foo(SOMETHING *obj, int *value)
{
    /* In OSSL_PARAM terms, this would be OSSL_PARAM_set_int(p, obj->foo)
       'cause you're setting |*value| to the current value of obj->foo */
    *value = obj->foo;
}
```

Reply:
I admit my mistake and will fix.

I also think this makes things even more confusing than they already were.  An alternative to the "set/get" wording would be good on _one_ side.  "push/pull" is the only thing that comes to mind immediately but they aren't great.

Reply:
"send/retrieve" ?

Reply:
@mattcaswell, I'd like to hear your thoughts on this as well.

Reply:
@levitte your explanation https://github.com/openssl/openssl/pull/9135#discussion_r294625745 is very clear, and I don't see any contradiction in it: `SOMETHING_set_foo` is a _setter_ of "class" `SOMETHING`, while `OSSL_PARAM_get_int` is a _getter_ of "class" `OSSL_PARAM`. Isn't it natural that you need to read the rhs value (using a getter) in order to assign it to the lhs (using a setter)?  Adding a new name pair ("push/pull", "set/respond") for the same concept doesn't clarify anything, it just makes things more complicated. Or did I miss some detail?

Reply:
Good point.

Reply:
I've updates the set functions to take const arrays of OSSL_PARAM again.